### ISSUE TITLE
`ed448-goldilocks`: Tracking Issue

### ISSUE THREAD

**Issue by daxpedda** [RustCrypto/elliptic-curves#1348]
This issue tracks various fixes, improvements and such for the `ed448-goldilocks` so we don't loose sight of them.

- [ ] Bugs
  - [ ] #1328
  - [ ] #1347
  - [ ] #1336
  - [ ] Remove torsion-check from `EdwardsPoint` de-serialization. These points are valid, even if not "ideal".
  - [ ] `DecafPoint` and `EdwardsPoint` de-serialization shouldn't allow invalid y-coordinates by just reducing them. This bug also introduces bias in `DecafPoint::try_from_rng()`.
  - [ ] Conversions between points of different curve types should account for identity points and other exceptions. We need tests for all of those cases.
  - [ ] #1326 (I didn't look into the EdDSA code at all yet)
- [ ] Optimizations
  - [ ] #1301
  - [ ] #1303
  - [ ] #1308
  - [ ] #1309
  - [ ] #1313
  - [ ] #1314
  - [ ] #1316
  - [ ] #1329
  - [ ] #1330
  - [ ] #1332
  - [ ] #1333
  - [ ] #1335
  - [ ] #1337
  - [ ] Open tracking issue for implementing ["Faster Complete Addition Laws for MontgomeryCurves"](https://tches.iacr.org/index.php/TCHES/article/view/11808/11313).
  - [ ] Add de/serialization for `DecafAffinePoint`. Here we can add `DecafAffinePoint::try_from_rng()` as well.
  - [ ] Add optimized algorithms for various operations in affine form for all curves.
  - [ ] Add optimized implementation for `LinearCombination` for all curves.
- [ ] Additions
  - [ ] #1306
  - [ ] #1334
  - [ ] Add `BatchNormalize` for `ProjectiveMontgomeryPoint` (#1306).
  - [ ] Add a checked `MontgomeryXpoint` de-serialization. The current one is unchecked because that's what X448 works.
- [ ] Align with other RustCrypto curve APIs
  - [ ] Remove methods covered by traits.
  - [ ] Remove `Compressed*` types in favor of `FieldBytes`.
  - [ ] Remove `(Wide)ScalarBytes` typedefs.
  - [ ] Use rejection sampling for `Scalar::try_from_rng()` instead of modular reduction.
  - [ ] Implement missing traits on affine forms (and `DecafPoint`): `DecompactPoint`, `DecompressPoint`, `GroupEncoding`.
  - [ ] Implement `serde::De/Serialize` for all affine forms (and `DecafPoint`).

**Comment by tarcieri** [comment]
Refactoring suggestion: move `src/curve/twedwards[.rs]` => `src/edwards/twisted[.rs]`

**Comment by daxpedda** [comment]
Decaf448 doesn't support batch encoding in any meaningful way, as opposed to Weierstrass curves where we can use batch normalization to make encoding cheaper. However, it does support double-and-compress batching. In practice, when you want to batch encode points that are a result of scalar multiplication, you can multiply by half the scalar and then double-and-compress batch the resulting points.

So I was thinking of introducing a new trait like `BatchMultiplyEncode`. This trait could be implemented on Weierstrass curves as well so protocols can rely on this trait without e.g. requiring specialization or some alternative workaround.

The same applies to Ristretto255.

Will open an issue in RustCrypto/traits after we implement `DecafPoint::batch_double_and_compress()` and I can show off some measurements.
